File: tests/algorithms/x/PHP/maths/find_max.php
<?php
ini_set('memory_limit', '-1');
$now_seed = 0;
$now_seeded = false;
$s = getenv('MOCHI_NOW_SEED');
if ($s !== false && $s !== '') {
    $now_seed = intval($s);
    $now_seeded = true;
}
function _now() {
    global $now_seed, $now_seeded;
    if ($now_seeded) {
        $now_seed = ($now_seed * 1664525 + 1013904223) % 2147483647;
        return $now_seed;
    }
    return hrtime(true);
}
function _intdiv($a, $b) {
    if (function_exists('bcdiv')) {
        $sa = is_int($a) ? strval($a) : (is_string($a) ? $a : sprintf('%.0f', $a));
        $sb = is_int($b) ? strval($b) : (is_string($b) ? $b : sprintf('%.0f', $b));
        return intval(bcdiv($sa, $sb, 0));
    }
    return intdiv($a, $b);
}
$__start_mem = memory_get_usage();
$__start = _now();
  function normalize_index($index, $n) {
  if ($index < 0) {
  return $n + $index;
}
  return $index;
};
  function find_max_iterative($nums) {
  if (count($nums) == 0) {
  $panic('find_max_iterative() arg is an empty sequence');
}
  $max_num = $nums[0];
  $i = 0;
  while ($i < count($nums)) {
  $x = $nums[$i];
  if ($x > $max_num) {
  $max_num = $x;
}
  $i = $i + 1;
};
  return $max_num;
};
  function find_max_recursive($nums, $left, $right) {
  $n = count($nums);
  if ($n == 0) {
  $panic('find_max_recursive() arg is an empty sequence');
}
  if ($left >= $n || $left < (0 - $n) || $right >= $n || $right < (0 - $n)) {
  $panic('list index out of range');
}
  $l = normalize_index($left, $n);
  $r = normalize_index($right, $n);
  if ($l == $r) {
  return $nums[$l];
}
  $mid = _intdiv(($l + $r), 2);
  $left_max = find_max_recursive($nums, $l, $mid);
  $right_max = find_max_recursive($nums, $mid + 1, $r);
  if ($left_max >= $right_max) {
  return $left_max;
}
  return $right_max;
};
  function test_find_max() {
  $arr = [2.0, 4.0, 9.0, 7.0, 19.0, 94.0, 5.0];
  if (find_max_iterative($arr) != 94.0) {
  $panic('find_max_iterative failed');
}
  if (find_max_recursive($arr, 0, count($arr) - 1) != 94.0) {
  $panic('find_max_recursive failed');
}
  if (find_max_recursive($arr, -count($arr), -1) != 94.0) {
  $panic('negative index handling failed');
}
};
  function main() {
  test_find_max();
  $nums = [2.0, 4.0, 9.0, 7.0, 19.0, 94.0, 5.0];
  echo rtrim(json_encode(find_max_iterative($nums), 1344)), PHP_EOL;
  echo rtrim(json_encode(find_max_recursive($nums, 0, count($nums) - 1), 1344)), PHP_EOL;
};
  main();
$__end = _now();
$__end_mem = memory_get_peak_usage();
$__duration = max(1, intdiv($__end - $__start, 1000));
$__mem_diff = max(0, $__end_mem - $__start_mem);
$__bench = ["duration_us" => $__duration, "memory_bytes" => $__mem_diff, "name" => "main"];
$__j = json_encode($__bench, 128);
$__j = str_replace("    ", "  ", $__j);
echo $__j, PHP_EOL;

File: tests/algorithms/transpiler/FS/maths/find_max.fs
// Generated 2025-08-16 14:41 +0700

exception Return
let mutable _nowSeed:int64 = 0L
let mutable _nowSeeded = false
let _initNow () =
    let s = System.Environment.GetEnvironmentVariable("MOCHI_NOW_SEED")
    if System.String.IsNullOrEmpty(s) |> not then
        match System.Int32.TryParse(s) with
        | true, v ->
            _nowSeed <- int64 v
            _nowSeeded <- true
        | _ -> ()
let _now () =
    if _nowSeeded then
        _nowSeed <- (_nowSeed * 1664525L + 1013904223L) % 2147483647L
        int _nowSeed
    else
        int (System.DateTime.UtcNow.Ticks % 2147483647L)

_initNow()
let _idx (arr:'a array) (i:int) : 'a =
    if not (obj.ReferenceEquals(arr, null)) && i >= 0 && i < arr.Length then arr.[i] else Unchecked.defaultof<'a>
let rec _str v =
    match box v with
    | :? float as f -> sprintf "%.15g" f
    | _ ->
        let s = sprintf "%A" v
        s.Replace("[|", "[")
         .Replace("|]", "]")
         .Replace("; ", " ")
         .Replace(";", "")
         .Replace("\"", "")
let _floordiv (a:int) (b:int) : int =
    let q = a / b
    let r = a % b
    if r <> 0 && ((a < 0) <> (b < 0)) then q - 1 else q
let rec normalize_index (index: int) (n: int) =
    let mutable __ret : int = Unchecked.defaultof<int>
    let mutable index = index
    let mutable n = n
    try
        __ret <- if index < 0 then (n + index) else index
        raise Return
        __ret
    with
        | Return -> __ret
and find_max_iterative (nums: float array) =
    let mutable __ret : float = Unchecked.defaultof<float>
    let mutable nums = nums
    try
        if (Seq.length (nums)) = 0 then
            ignore (failwith ("find_max_iterative() arg is an empty sequence"))
        let mutable max_num: float = _idx nums (int 0)
        let mutable i: int = 0
        while i < (Seq.length (nums)) do
            let x: float = _idx nums (int i)
            if x > max_num then
                max_num <- x
            i <- i + 1
        __ret <- max_num
        raise Return
        __ret
    with
        | Return -> __ret
and find_max_recursive (nums: float array) (left: int) (right: int) =
    let mutable __ret : float = Unchecked.defaultof<float>
    let mutable nums = nums
    let mutable left = left
    let mutable right = right
    try
        let n: int = Seq.length (nums)
        if n = 0 then
            ignore (failwith ("find_max_recursive() arg is an empty sequence"))
        if (((left >= n) || (left < (0 - n))) || (right >= n)) || (right < (0 - n)) then
            ignore (failwith ("list index out of range"))
        let mutable l: int = normalize_index (left) (n)
        let mutable r: int = normalize_index (right) (n)
        if l = r then
            __ret <- _idx nums (int l)
            raise Return
        let mid: int = _floordiv (int (l + r)) (int 2)
        let left_max: float = find_max_recursive (nums) (l) (mid)
        let right_max: float = find_max_recursive (nums) (mid + 1) (r)
        if left_max >= right_max then
            __ret <- left_max
            raise Return
        __ret <- right_max
        raise Return
        __ret
    with
        | Return -> __ret
and test_find_max () =
    let mutable __ret : obj = Unchecked.defaultof<obj>
    try
        let arr: float array = unbox<float array> [|2.0; 4.0; 9.0; 7.0; 19.0; 94.0; 5.0|]
        if (find_max_iterative (arr)) <> 94.0 then
            ignore (failwith ("find_max_iterative failed"))
        if (find_max_recursive (arr) (0) ((Seq.length (arr)) - 1)) <> 94.0 then
            ignore (failwith ("find_max_recursive failed"))
        if (find_max_recursive (arr) (-(Seq.length (arr))) (-1)) <> 94.0 then
            ignore (failwith ("negative index handling failed"))
        __ret
    with
        | Return -> __ret
and main () =
    let mutable __ret : obj = Unchecked.defaultof<obj>
    try
        let __bench_start = _now()
        let __mem_start = System.GC.GetTotalMemory(true)
        ignore (test_find_max())
        let nums: float array = unbox<float array> [|2.0; 4.0; 9.0; 7.0; 19.0; 94.0; 5.0|]
        ignore (printfn "%s" (_str (find_max_iterative (nums))))
        ignore (printfn "%s" (_str (find_max_recursive (nums) (0) ((Seq.length (nums)) - 1))))
        let __bench_end = _now()
        let __mem_end = System.GC.GetTotalMemory(true)
        printfn "{\n  \"duration_us\": %d,\n  \"memory_bytes\": %d,\n  \"name\": \"main\"\n}" ((__bench_end - __bench_start) / 1000) (__mem_end - __mem_start)

        __ret
    with
        | Return -> __ret
ignore (main())
